### PR TITLE
RFC: Change 'readdir' to return an iterator.

### PR DESCRIPTION
**Pull request by malmaud**
cf https://github.com/JuliaLang/julia/issues/27393

### CODE REVIEW

File: base/file.jl [hunk 1]
```diff
@@ -572,14 +572,41 @@ struct uv_dirent_t
     typ::Cint
 end
 
+struct ReadDirIter
```
Comment (StefanKarpinski) [inline base/file.jl]: This seems unnecessarily abbreviated. How about `ReadDirIterator`?

**Review by StefanKarpinski — APPROVED** [review]
Aside from iterator name bikeshedding, looks good to me.

**Comment by vtjnash** [comment]
It would be simpler (and faster and more memory efficient) to implement this as:

```julia
lazy_readdir(name) = (path for path in readdir(name))
```

Just my 2¢

**Comment by malmaud** [comment]
Wait, how is that lazy? It would still be loading all the paths into memory
at once, wouldnt it?

**Comment by vtjnash** [comment]
They are already all loaded into memory, this PR would just make accessing them slower.

Support for the lazy version was proposed a few years ago, so it might land soon(ish) in libuv master: https://github.com/libuv/libuv/pull/416

**Comment by StefanKarpinski** [comment]
I guess the question for us at this point is: do we want to expose an iterator interface to `readdir` even though the current `libuv` implementation is always eager in anticipation of `libuv` soon having a truly lazy implementation for which an iterator interface can be advantageous. I would say yes. The choice of implementation doesn't matter much and @malmaud's one here has the advantage of possibly not needing to change when `libuv` changes theirs.

**Comment by malmaud** [comment]
I agree with that. Ill also run a quick benchmark just in case there really
is a non trivial performance hit.

File: base/file.jl (hunk 1 above)
Comment (malmaud) [reply base/file.jl]: Sure

**Comment by JeffBezanson** [comment]
I don't see why an iterator is better. Arrays support more operations, e.g. `pmap`. Directories hardly ever have so many entries that memory use would be prohibitive.

**Comment by StefanKarpinski** [comment]
The fact that it's an issue for libuv would seem to indicate that there are cases where it matters. Otherwise they would presumably not be bothering to change this.

**Comment by StefanKarpinski** [comment]
Relevant motivational issues:
* https://github.com/nodejs/node/issues/15699
* https://github.com/nodejs/node-v0.x-archive/issues/388

**Comment by malmaud** [comment]
Yes, I agree that the fact that most other libraries already do this lazily or have open issues from people running into trouble from it not being lazy should tell us something.  

I think there's no practical limit in Linux anymore on how many files can be in a directory.  Thus someone, somewhere,  will run into performance issues that we can never remedy once we've committed to an API where `readdir` returns a fully reified list.

We could potentially just add another function like Jameson's `lazy_readdir` or whatever, if people want to keep `readdir` as it is.

**Comment by JeffBezanson** [comment]
Ok, I'm convinced, we can make it an iterator.

**Comment by malmaud** [comment]
Big mistake.

**Comment by StefanKarpinski** [comment]
Note that to address the "double file system call" issue in https://github.com/nodejs/node/issues/15699 we would need to return some kind of `FileNode` object that acts like a file path but caches stat information (which it already has from the `readdir` operation). It's unclear to me how relevant the "if we list large directories the whole process hangs" issue in https://github.com/nodejs/node/issues/15699 is to Julia—after all, we can make `readdir` internally non-blocking and let other tasks do stuff while `readdir` collects directory names. Blocking isn't the huge problem in Julia that it is in JavaScript.

**Comment by StefanKarpinski** [comment]
Bump, it's now or never. This is unlikely to break _too_ terribly many usages since the most common usage is simply to iterate through the output of `readdir`...

**Comment by vtjnash** [comment]
It's still not clear to me how often you would need to use this function on directories with more than tens of millions of entries, such that you would want to slightly increase the cost and complexity of using `readdir`.

**Comment by StefanKarpinski** [comment]
If that's the call, ok. If it becomes an issue in the future, we could even have `readdir(dir, lazy=true)` or something like it in the future.

**Comment by malmaud** [comment]
Alright, shall we just do Stefan's suggestion then (presumably a type-stable variant) at some point then?

**Pull request closed without merging**